### PR TITLE
Remove storage layer abstractions

### PR DESCRIPTION
**Pull request by seaneganx**
This PR removes all remaining storage layer abstractions that are no longer useful after #130. Specifically the `Query`, `Iterator`, `Key`, and `Requirement` types were removed, and the `gorm` package was merged with the `storage` package.

My goal with this is to simplify our implementation and decouple each database method as much as possible before addressing other database issues like #148, #257, #382, and #414. 

The file diffs aren't going to be very useful in reviewing this PR because files have been modified and moved, so many unchanged sections of the package are going to appear as new additions. The list of commits might be helpful.

### CODE REVIEW

**Review by timburks — APPROVED** [review]
Nice work! Breaking this up into small CLs made this much easier to review. I'm curious to know how much you planned these steps in advance.

I had one functional concern that I manually verified - perhaps this is also represented in tests. I wanted to be sure that filtering still worked on large collections of resources when the requested items were sparse and possibly at the very end of the list. (I checked and this was fine.) On reading the commits, I see no reason why it wouldn't have worked, since the filtered list handlers always get everything in one large page of results.

File: server/registry/internal/storage/gorm/list.go [hunk 1]
```diff
@@ -57,10 +58,16 @@ func (c *Client) ListProjects(ctx context.Context, opts PageOptions) (ProjectLis
...
 	}
 
-	q := c.NewQuery(ProjectEntityName)
-	q = q.ApplyOffset(token.Offset)
+	lock()
+	var v []models.Project
+	_ = c.db.
+		Order("key").
+		Offset(token.Offset).
+		Limit(100000).
```
[the remark this answers is not in the document]
Comment (seaneganx) [reply server/registry/internal/storage/gorm/list.go]: I'm curious to see how our system performs as we fill our database incrementally up to 100k rows per table. Of course >100k we can break filtering but there might be a point much earlier where memory issues begin to cause failures.

I agree we might reintroduce an iterator of some kind. It should be easier to introduce that incrementally from this blank(er) slate than modify it while all of our listing methods depend on its current API. I gave a quick attempt at addressing #148 before making these changes and felt like [Hyrum's Law](https://www.hyrumslaw.com/) was working against me.

File: server/registry/internal/storage/gorm/get.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+	"github.com/apigee/registry/server/registry/internal/storage/models"
+	"github.com/apigee/registry/server/registry/names"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"gorm.io/gorm"
+)
+
+func (c *Client) GetProject(ctx context.Context, name names.Project) (*models.Project, error) {
+	v := new(models.Project)
```
[the remark this answers is not in the document]
Comment (seaneganx) [reply server/registry/internal/storage/gorm/get.go]: I don't think this changed, at least not for the Get methods. Those are interchangeable syntax and I don't think there's any common guidance on when to use one or the other.

**Comment by seaneganx** [comment]
>Nice work! Breaking this up into small commits made this much easier to review. I'm curious to know how much you planned these steps in advance.

I didn't plan all the individual changes in advance (just the end goal). I did plan to use small enough commits so that each is individually coherent because I knew the full diff would be a mess.

>I had one functional concern that I manually verified - perhaps this is also represented in tests. I wanted to be sure that filtering still worked on large collections of resources when the requested items were sparse and possibly at the very end of the list. (I checked and this was fine.) On reading the commits, I see no reason why it wouldn't have worked, since the filtered list handlers always get everything in one large page of results.

We have tests ([example](https://github.com/apigee/registry/blob/c9b8f611da51823adc688374815ac335db2c0b3f/server/registry/actions_apis_test.go#L693)) that make sure pagination works correctly with filters when the matched elements are sparse, but it only uses 100 resources. Does that cover everything you're thinking of?

**Pull request merged**